File: CSharpAthletic/NetworkProto/ProtoWorker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Threading;
using Google.Protobuf;
using Services;

namespace NetworkProto
{
    public class ProtoWorker : IObserver
    {
        private readonly IServices Server;
        private readonly TcpClient Connection;
        private readonly NetworkStream Stream;
        private volatile bool Connected;

        public ProtoWorker(IServices server, TcpClient connection)
        {
            Server = server;
            Connection = connection;
            try
            {
                Stream = connection.GetStream();
                Connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public virtual void RefreshEvents(IEnumerable<Domain.DTO.EventCountDTO> eventCountList)
        {
            try
            {
                Console.WriteLine("Refresh Event Count List Response");
                Domain.DTO.EventCountDTO[] lst = eventCountList.ToArray();
                SendResponse(ProtoUtils.GetEventsNumberResponse(lst, true));
            }
            catch (Exception e)
            {
                throw new Exception(e.StackTrace);
            }
        }

        public virtual void Run()
        {
            while (Connected)
            {
                try
                {
                    Request request = Request.Parser.ParseDelimitedFrom(Stream);
                    Response response = HandleRequest(request);
                    if (response != null)
                    {
                        SendResponse(response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                Stream.Close();
                Connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private Response HandleRequest(Request request)
        {
            if (request.Type is Request.Types.Type.Login)
            {
                Console.WriteLine("Login request");
                Domain.User user = ProtoUtils.GetUser(request);
                try
                {
                    lock (Server)
                    {
                        Server.Login(user, this);
                    }
                    return ProtoUtils.OkResponse();
                }
                catch (Exception e)
                {
                    Connected = false;
                    return ProtoUtils.ErrorResponse(e.Message);
                }
            }

            if (request.Type is Request.Types.Type.Logout)
            {
                Console.WriteLine("Logout request");
                Domain.User user = ProtoUtils.GetUser(request);
                try
                {
                    lock (Server)
                    {
                        Server.Logout(user, this);
                    }
                    Connected = false;
                    return ProtoUtils.OkResponse();
                }
                catch (Exception e)
                {
                    return ProtoUtils.ErrorResponse(e.Message);
                }
            }

            if (request.Type is Request.Types.Type.ChildrenEvents)
            {
                Console.WriteLine("Children Events Request");
                Domain.DTO.EventAgeGroupDTO eag = ProtoUtils.GetEventAgeGroupDTO(request);
                try
                {
                    Domain.DTO.ChildNoEventsDTO[] lst;
                    lock (Server)
                    {
                        lst = Server.GetChildrenNoEvents(eag.Event, eag.AgeGroup).ToArray();
                    }
                    return ProtoUtils.GetChildrenNoEventsResponse(lst);
                }
                catch (Exception e)
                {
                    return ProtoUtils.ErrorResponse(e.Message);
                }
            }

            if (request.Type is Request.Types.Type.AddParticipant)
            {
                Console.WriteLine("Add Participant Request");
                Domain.DTO.ParticipantDetailsDTO pd = ProtoUtils.GetParticipantDetailsDTO(request);
                try
                {
                    lock (Server)
                    {
                        Server.AddParticipant(pd.Name, pd.Age, pd.Event1, pd.Event2);
                    }
                    return ProtoUtils.OkResponse();
                }
                catch (Exception e)
                {
                    return ProtoUtils.ErrorResponse(e.Message);
                }
            }

            if (request.Type is Request.Types.Type.EventCount)
            {
                Console.WriteLine("Event Count Request");
                try
                {
                    Domain.DTO.EventCountDTO[] lst;
                    lock (Server)
                    {
                        lst = Server.GetEventsNumber().ToArray();
                    }
                    return ProtoUtils.GetEventsNumberResponse(lst, false);
                }
                catch (Exception e)
                {
                    return ProtoUtils.ErrorResponse(e.Message);
                }
            }
            return null;
        }

        private void SendResponse(Response response)
        {
            lock (Stream)
            {
                response.WriteDelimitedTo(Stream);
                Stream.Flush();
            }
        }
    }
}

File: JavaAthletic/AthleticNetwork/src/main/java/ProtoWorker.java
import domain.User;
import dto.ParticipantDetailsDTO;
import dtos.ChildNoEventsDTO;
import dtos.EventAgeGroupDTO;
import dtos.EventCountDTO;

import java.io.*;
import java.net.Socket;
import java.util.List;

public class ProtoWorker implements Runnable, IObserver {
    private final IServiceServer server;
    private final Socket connection;

    private InputStream input;
    private OutputStream output;
    private volatile boolean connected;

    public ProtoWorker(IServiceServer server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try {
            output = connection.getOutputStream();
            input = connection.getInputStream();
            connected = true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }


    @Override
    public void refreshEvents(List<EventCountDTO> eventCountList) throws Exception {
        EventCountDTO[] arr = eventCountList.toArray(new EventCountDTO[0]);
        try {
            sendResponse(ProtoUtils.getEventsNumberResponse(arr, true));
        }
        catch (IOException e) {
            throw new Exception("Sending error\n" + e);
        }
    }

    @Override
    public void run() {
        while(connected) {
            try {
                Protobuf.Request request = Protobuf.Request.parseDelimitedFrom(input);
                Protobuf.Response response = handleRequest(request);
                if (response != null) {
                    sendResponse(response);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("Error\n" + e);
        }
    }

    private void sendResponse(Protobuf.Response response) throws IOException {
        System.out.println("Sending response " + response);
        response.writeDelimitedTo(output);
        output.flush();
    }

    private Protobuf.Response handleRequest(Protobuf.Request request) {
        if (request.getType() == Protobuf.Request.Type.LOGIN) {
            System.out.println("Login request");
            User user = ProtoUtils.getUser(request);
            try {
                server.login(user, this);
                return ProtoUtils.okResponse();
            } catch (Exception e) {
                connected = false;
                return ProtoUtils.errorResponse(e.getMessage());
            }
        }
        if (request.getType() == Protobuf.Request.Type.LOGOUT) {
            System.out.println("Logout request");
            User user = ProtoUtils.getUser(request);
            try {
                server.logout(user, this);
                connected = false;
                return ProtoUtils.okResponse();
            } catch (Exception e) {
                return ProtoUtils.errorResponse(e.getMessage());
            }
        }
        if (request.getType() == Protobuf.Request.Type.CHILDREN_EVENTS) {
            System.out.println("Children events request");
            EventAgeGroupDTO eag = ProtoUtils.getEventAgeGroupDTO(request);
            try {
                List<ChildNoEventsDTO> lst = server.getChildrenNoEvents(eag.getEvent(), eag.getAgeGroup());
                ChildNoEventsDTO[] arr = lst.toArray(new ChildNoEventsDTO[0]);
                return ProtoUtils.getChildrenNoEventsResponse(arr);
            }
            catch (Exception e) {
                return ProtoUtils.errorResponse(e.getMessage());
            }
        }
        if (request.getType() == Protobuf.Request.Type.ADD_PARTICIPANT) {
            System.out.println("Add participant request");
            ParticipantDetailsDTO pd = ProtoUtils.getParticipantDetailsDTO(request);
            try {
                server.addParticipant(pd.getName(), pd.getAge(), pd.getEvent1(), pd.getEvent2());
                return ProtoUtils.okResponse();
            } catch (Exception e) {
                return ProtoUtils.errorResponse(e.getMessage());
            }
        }
        if (request.getType() == Protobuf.Request.Type.EVENT_COUNT) {
            System.out.println("Event count request");
            try {
                List<EventCountDTO> lst = server.getEventsNumber();
                EventCountDTO[] arr = lst.toArray(new EventCountDTO[0]);
                return ProtoUtils.getEventsNumberResponse(arr, false);

            } catch (Exception e) {
                return ProtoUtils.errorResponse(e.getMessage());
            }
        }
        return null;
    }
}
